src/hotspot/share/gc/g1/g1ParScanThreadState.inline.hpp

```diff
@@ -77,7 +77,11 @@ inline void G1ParScanThreadState::do_oop_partial_array(oop* p) {
   oop from_obj = clear_partial_array_mask(p);
 
   assert(_g1h->is_in_reserved(from_obj), "must be in heap.");
-  assert(from_obj->is_objArray(), "must be obj array");
+  if (UseCompactObjectHeaders) {
+    assert(from_obj->forward_safe_klass()->is_objArray_klass(), "must be obj array");
+  } else {
+    assert(from_obj->is_objArray(), "must be obj array");
+  }
```

Comment:
Looks like we can always use the first form (line 81)

Reply:
just want to keep the unchanged branch